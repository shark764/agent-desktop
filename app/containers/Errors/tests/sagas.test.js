import Raven from 'raven-js';

import {
  goHandleSDKError,
  topicActions,
} from 'containers/Errors/sagas';

jest.mock('raven-js', () => ({
  captureException: jest.fn(),
}));

describe('handleError Saga', () => {
  let generator;
  const mockAction = {
    error: {
      code: 'mockErrorCode',
    },
    topic: 'mockErrorTopic',
  };
  beforeEach(() => {
    Raven.captureException.mockClear();
    generator = goHandleSDKError(mockAction);
    generator.next();
  });
  it('should call Raven.captureException', () => {
    expect(Raven.captureException.mock.calls.length).toBe(1);
  });
  it('should call Raven.captureException with a new Error created from the action.topic', () => {
    const result = Raven.captureException.mock.calls[0][0];
    delete result.stack; // don't want to mess with Error and don't want diff snapshots everytime
    expect(result).toMatchSnapshot();
  });
  it('should call Raven.captureException with the SDK error in the extra info and error.code in the tags', () => {
    expect(Raven.captureException.mock.calls[0][1]).toMatchSnapshot();
  });

  Object.keys(topicActions).forEach((actionTopic) => {
    describe(`if topic is ${actionTopic}`, () => {
      beforeEach(() => {
        mockAction.topic = actionTopic;
        generator = goHandleSDKError(mockAction);
      });
      it(`should dispatch action ${topicActions[actionTopic].type}`, () => {
        expect(generator.next()).toMatchSnapshot();
      });
    });
  });

  describe('if error.level is "fatal"', () => {
    beforeEach(() => {
      mockAction.error = {
        level: 'fatal',
      };
      generator = goHandleSDKError(mockAction);
    });
    it('should dispatch action setCriticalError', () => {
      expect(generator.next()).toMatchSnapshot();
    });
    it('should call Raven.captureException with level fatal', () => {
      generator.next();
      const result = Raven.captureException.mock.calls[0][1];
      expect(result.level).toBe('fatal');
    });
  });
});
