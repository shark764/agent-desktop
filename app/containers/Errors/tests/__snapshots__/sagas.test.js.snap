// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleError Saga if error.level is "fatal" should dispatch action setCriticalError 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "type": "app/Errors/SET_CRITICAL_ERROR",
      },
      "channel": null,
    },
  },
}
`;

exports[`handleError Saga if topic is cxengage/contacts/list-attributes-response should dispatch action app/InfoTab/SET_CRM_UNAVAILABLE 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "reason": "crmAttributeError",
        "type": "app/InfoTab/SET_CRM_UNAVAILABLE",
      },
      "channel": null,
    },
  },
}
`;

exports[`handleError Saga if topic is cxengage/contacts/list-layouts-response should dispatch action app/InfoTab/SET_CRM_UNAVAILABLE 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "reason": "crmLayoutError",
        "type": "app/InfoTab/SET_CRM_UNAVAILABLE",
      },
      "channel": null,
    },
  },
}
`;

exports[`handleError Saga if topic is cxengage/session/heartbeat-response should dispatch action setCriticalError 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "type": "app/Errors/SET_CRITICAL_ERROR",
      },
      "channel": null,
    },
  },
}
`;

exports[`handleError Saga should call Raven.captureException with a new Error created from the action.topic 1`] = `[Error: mockErrorTopic]`;

exports[`handleError Saga should call Raven.captureException with the SDK error in the extra info and error.code in the tags 1`] = `
Object {
  "extra": Object {
    "sdkError": Object {
      "code": "mockErrorCode",
    },
  },
  "level": "warning",
  "tags": Object {
    "sdkCode": "mockErrorCode",
  },
}
`;
